Add standard trigger for separation between objects

knelli2 opened the pull request:
## Proposed changes

Checks the distance between `domain::Tags::ObjectCenter<A>` and `B` in the inertial. If it's less than the input, then it's triggered.

This is a rough start to finding common horizons (or a horizon for BNS). It will certainly need to be modified/updated in the future (probably to a DenseTrigger even)

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add standard trigger for separation between objects

wthrowe commented on src/Domain/Triggers.hpp:
```diff
@@ -0,0 +1,98 @@
+      "Trigger when the separation between the two horizons is less than a "
+      "certain distance."};
+
+  SeparationLessThan(double separation);
```
`explicit`

wthrowe commented on src/Domain/Triggers.cpp:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  // reuse allocation
+  tnsr::I<double, 3, Frame::Inertial>& position_difference =
```
There is no allocation.

wthrowe commented on src/Domain/Triggers.cpp:
```diff
@@ -0,0 +1,81 @@
+  tnsr::I<double, 3, Frame::Inertial>& position_difference =
+      inertial_object_center_a;
+
+  position_difference = tenex::evaluate<ti::I>(inertial_object_center_a(ti::I) -
```
If there were an allocation, you would have to use the by-reference version to reuse it.  (But there isn't so this is fine.)

wthrowe commented on src/Domain/Triggers.cpp:
```diff
@@ -0,0 +1,81 @@
+  const ElementId<3>& element_id = element.id();
+  const size_t block_id = element_id.block_id();
+
+  const Block<3>& block = domain.blocks()[block_id];
```
I'm confused about what's going on here.  It looks like you are using the maps from the current block to map the object centers.  Are these maps guaranteed to be the same for every element?

knelli2 replied:
So I wasn't sure how to do this because we literally don't have elements at the centers of the horizons. So I can't check if the center is in an element/block to specifically use that map. But this trigger needs to be run on all elements. My solution was to make use of the fact that currently, all Grid->Distorted maps leave the centers of the objects unchanged, and all Distorted->Inertial maps are global. So if a block didn't have a distorted frame, I just call the Grid->Inertial map because I know they are (currently) defined everywhere. And if a block has a distorted frame, I just call the Distorted->Inertial map which (currently) should be the same as the Grid->Inertial map without a distorted frame.

This is all very hacky. But I couldn't think of a better way. Ideas are welcome

wthrowe replied:
OK, it looks like that's true in SpEC, so it should be fine here, at least for now.

knelli2 replied:
Yeah while we have Shape/Size/Rotaiton/Expansion/Translation, this will be true. Once we add Skew/Cut-X it won't be and we'll need to find a different way. Possibly we add some kind of thing that holds coordinate maps for inside the excision, except it doesn't actually have grid points/solutions.

But for now, I'll leave it as is.

wthrowe replied:
Oh, you're right.  I misread the SpEC paper.  How does SpEC do this?

knelli2 replied:
I honestly don't know. @markscheel?

nilsdeppe replied:
SpEC's maps are all global and then if-else branches depending on coordinate locations inside the calls

knelli2 replied:
Ah ok yeah that explains it. I'm thinking that we just have to imbue some object with coordinate maps that are valid inside the excision spheres. Maybe the `ExcisionSphere`s?

nilsvu replied:
Yea I think adding coord maps to excision spheres would be reasonable. I think @nikwit needs something like that as well for worldtube things.

wthrowe replied:
I agree, that makes sense.  Excision spheres are kind of like blocks in that they are covering a section of the grid-coordinates, just not a section with elements, so having some similar properties is reasonable.

knelli2 replied:
Ok I'll add maps to the `ExcisionSphere`s (probably in a separate PR).

wthrowe replied:
Sounds good.  I don't think we need to hold this PR up for it.  Just fix it in the PR that adds the feature.  If you're not going to do it immediately file a bug so we don't forget.

wthrowe commented on src/Domain/Triggers.hpp:
```diff
@@ -0,0 +1,98 @@
+#include "Domain/Structure/ObjectLabel.hpp"
+#include "Options/String.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Trigger.hpp"
+#include "Time/Utilities.hpp"
```
You don't use this.

wthrowe commented on src/Domain/Triggers.hpp:
```diff
@@ -0,0 +1,98 @@
+// Distributed under the MIT License.
```
I think this should be in a less generically named file.

nilsvu replied:
Maybe `Evolution/Triggers/SeparationLessThan`?

wthrowe replied:
Yes, let's get it out of Domain to avoid a new dependency there.

knelli2 commented:
Made (and squashed) all the changes other than the new excision sphere stuff which I'll do in a new PR. Added this trigger to a new `EvolutionTriggers` library.

wthrowe commented on src/Evolution/Triggers/SeparationLessThan.cpp:
```diff
@@ -0,0 +1,79 @@
+        grid_to_inertial_map(object_center_b, time, functions_of_time);
+  }
+
+  // reuse allocation
```
Remove comment.  Still not true.

nilsvu commented on src/Evolution/Triggers/SeparationLessThan.cpp:
```diff
@@ -0,0 +1,78 @@
+  tnsr::I<double, 3, Frame::Inertial> inertial_object_center_a{};
+  tnsr::I<double, 3, Frame::Inertial> inertial_object_center_b{};
+
+  if (block.has_distorted_frame()) {
```
I think you don't need this because the grid-to-inertial map is available even when you have a distorted frame. Please double-check.

knelli2 replied:
Yeah but the Shape map isn't global, so it isn't valid inside the excision sphere. Only expansion and rotation are. So in a block with a distorted frame, we can't call the shape map, so we have to call the distorted to inertial map. However, all of this is mute with #5184.

nilsvu replied:
Ah ok, this should definitely be in a comment. Doesn't matter though if you'll refactor it with #5184 anyway.
